src/nips/nip13.rs

```diff
@@ -0,0 +1,115 @@
+use crate::{
+    events::{Event, EventPrepare},
+    nostr_client::Client,
+    utils::{get_timestamp, self},
+    Identity,
+};
+use rand::Rng;
+use secp256k1::{KeyPair, SECP256K1};
+
+// Implementation of the NIP13 protocol
+// https://github.com/nostr-protocol/nips/blob/master/13.md
+
+impl EventPrepare {
+    pub fn count_leading_zero_bits(content_id: Vec<u8>) -> u32 {
+        let mut total: u32 = 0;
+
+        for c in content_id {
+            let bits = c.leading_zeros();
+            total += bits;
+            if bits != 8 {
+                break;
+            }
+        }
+        total
+    }
+
+    pub fn to_pow_event(
+        &mut self,
+        secret_key: &Identity,
+        difficulty: u32,
+    ) -> Result<Event, String> {
+        let mut rng = rand::thread_rng();
+        loop {
+            let nouce: u32 = rng.gen_range(0..999999);
+
+            self.tags.push(vec![
+                "nouce".to_string(),
+                nouce.to_string(),
+                difficulty.to_string(),
+            ]);
+
+            let content_id = self.get_content_id();
+            let content_id = hex::decode(content_id).map_err(|e| e.to_string())?;
+            if Self::count_leading_zero_bits(content_id) >= difficulty {
+                break;
+            }
+            self.tags.pop();
+            self.created_at = utils::get_timestamp();
+        }
+
+        let message = secp256k1::Message::from_hashed_data::<secp256k1::hashes::sha256::Hash>(
+            self.get_content().as_bytes(),
+        );
+
+        let signature = SECP256K1
+            .sign_schnorr(
+                &message,
+                &KeyPair::from_secret_key(SECP256K1, &secret_key.secret_key),
+            )
+            .to_string();
+
+        Ok(Event {
+            id: self.get_content_id(),
+            pub_key: self.pub_key.clone(),
+            created_at: self.created_at,
+            kind: self.kind,
+            tags: self.tags.clone(),
+            content: self.content.clone(),
+            sig: signature,
+        })
+    }
+}
+
+impl Client {
+    /// Publish a text note event with Proof of Work
+    /// # Example
+    /// ```rust
+    /// use nostr_rust::{nostr_client::Client, Identity, utils::get_timestamp};
+    /// use std::str::FromStr;
+    /// let mut client = Client::new(vec!["wss://nostr-pub.wellorder.net"]).unwrap();
+    /// let identity = Identity::from_str(env!("SECRET_KEY")).unwrap();
+    /// let message = format!("Hello Nostr! {}", get_timestamp());
+    /// client.publish_pow_text_note(&identity, &message, &vec![], 20).unwrap();
```

Comment:
How long does it take to calculate a difficulty of 20?
Because for the tests with Github Action, it would have to fall on 5 seconds max to not cause problems and I have no idea how much that is 20

Reply:
In the testing I've done difficulty less then 12 is consistently less then a second. 12-18 is about 1-5 seconds, 20 is usually 30-45 seconds. So for the github action we should lower the difficulty. 

But I'm not really sure how long it is supposed to take, to know if I'm on target or not and 25 has finished in 4min once but other times I've stopped it after 10+ min without it finishing, which leads me to think I've made a mistake somewhere. 
I will try to write a few benchmarks to get a better idea of the numbers

Reply:
Maybe you were very lucky once!
You should mine bitcoin haha

Reply:
```rust
for diff in [5, 10, 15, 20, 25] {
        let mut times = vec![];
        for _ in 0..10 {
            let now = time::Instant::now();
            let content = format!("Hello POW diff {}", diff);
            nostr_client
                .lock()
                .unwrap()
                .publish_pow_text_note(&my_identity, &content, &[], diff)
                .unwrap();
            let elapsed_time = now.elapsed();
            times.push(elapsed_time.as_millis());
        }

        let sum: u128 = times.iter().sum();
        let avg = sum as f32 / times.len() as f32;
        println!("Avg for difficulty {} is {} milliseconds", diff, avg);
    }
```
Avg for difficulty 5 is 0 milliseconds
Avg for difficulty 10 is 3.5 milliseconds
Avg for difficulty 15 is 70.8 milliseconds
Avg for difficulty 20 is 2891.2 milliseconds
Avg for difficulty 25 is 44998.4 milliseconds

Haven't been lucky enough to get one for difficulty 30 yet so I'll have to hold off on the bitcoin mining lol